include/deal.II/non_matching/quadrature_generator.h

```diff
@@ -0,0 +1,211 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_non_matching_quadrature_generator
+#define dealii_non_matching_quadrature_generator
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/point.h>
+
+#include <type_traits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  namespace internal
+  {
+    namespace QuadratureGeneratorImplementation
+    {
+      /**
+       * Describes a dim-dimensional hypercube aligned with the coordinate
+       * axes. That is, a region
+       *
+       * [x_i - L/2, x_i + L/2]^dim
+       *
+       * where L is the side length and (x_0, ..., x_{dim-1}) are the
+       * coordinates of the center of the hypercube.
+       *
+       * The purpose of this class is to help us keep track of where we are on
+       * the reference cell when generating immersed quadrature rules. There are
+       * two main things we want to be able to do.
+       *
+       * 1) Split a hypercube into its 2^dim children.
+       * 2) Take a cross section orthogonal to a given direction and get this as
+       *    a Hypercube<dim-1>.
+       *
+       * These are needed because the algorithm is allowed to recurse both over
+       * children and over dimensions.
+       *
+       * In 3D, the 2 coordinates of the cross section of HyperCube<3> can be
+       * ordered in 2 ways. That is, if we take the cross section orthogonal to
+       * the y direction we could either order a 3D-coordinate into a
+       * 2D-coordinate as (x,z) or as (z,x). This class uses the second
+       * convention, corresponding to the coordinates being ordered cyclicly
+       * x -> y -> z -> x -> ... To be precise, if we take a cross section:
```

Comment:
There is some weird doxygen stuff going on with the arrows and tables:
![Screenshot_2020-03-14 The deal II Library NonMatching internal QuadratureGeneratorImplementation Hypercube dim Class Templa](https://user-images.githubusercontent.com/3917035/76681746-b99c8d80-65cc-11ea-961b-28550f4e0ca0.png)